quinn-proto/src/connection.rs

```diff
@@ -1341,9 +1265,9 @@ impl Connection {
                         self.drive_tls(ctx, &mut state.tls)?;
                         Ok(x)
                     }) {
-                    Err(e) => State::closed(e),
-                    Ok(true) => State::Draining(state.into()),
-                    Ok(false) => State::Established(state),
+                    Err(e) => Ok(State::closed(e)),
+                    Ok(true) => Ok(State::Draining),
+                    Ok(false) => Ok(State::Established(state)),
                 }
```

Comment:
Could this whole block be rewritten to something like the following?
```rust
let closed = self.process_payload(ctx, now, number, payload.into(), &mut state.tls)?;
self.drive_tls(ctx, &mut state.tls)?;
Ok(if closed { State::Draining } else { State::Established(state) })
```
You'd want to remove the `ConnectionLost` events from `process_payload` if so, letting them be generated automatically like the rest.

Reply:
Using `and_then` still makes calling `drive_tls()` contingent on whether the result of `process_payload()` is `Ok()`, so you'd still need more pattern matching than what you have shown here.

Reply:
I don't follow. `self.process_payload(...)?` will cause the enclosing function to return immediately if the call doesn't return `Ok()`.

Reply:
Ah, you're right -- I missed that.